[TRTLLM-9465][fix] Swap TP-CP grouping order

brb-nv opened the pull request:
## Description
This MR fixes poor accuracy on GSM8K accuracy benchmark when Helix CP is used along with TP on gen server.

**_Background:_**
- When using helix parallelism, CP ranks are repurposed to TP ranks. This happens not just for the FFN part, but even for the `o_proj` part of attention layer.

**_Root cause:_**
- When there's a mix of TP & HelixCP (TP=2/CP=2 for example), all ranks become TP ranks for `o_proj` but there's an issue in the order of these ranks because CP ranks are interleaved.
- In above example, cp_groups are {0, **_2_**}, {**_1_**,3} but repurposed tp_group is {0,**_1,2_**,3}. Note the difference in order of ranks highlighted in bold. So, o_proj has the wrong weights on a subset of ranks.

**_Fix:_**
- Update the TP-CP grouping order such that when CP is present, CP ranks are consecutive while TP ranks are interleaved (with stride `cp_size`) before repurposing.
- After repurposing with cp_size=1, TP ranks are sequential anyway and `o_proj` will receive the right weights.
- Alternative fix considered: https://github.com/NVIDIA/TensorRT-LLM/pull/9759

## Test Coverage
```
$ pytest tests/unittest/others/test_mapping.py -s -v
$ TRTLLM_USE_UCX_KVCACHE=1 mpirun -n 8 ./tests/unit_tests/multi_gpu/cacheTransceiverTest --gtest_filter="AsymmetricCaseTest0WithCPForMLA/AsymmetricalCacheTest.TestCase/*"
$ TRTLLM_USE_UCX_KVCACHE=1 mpirun -n 8 ./tests/unit_tests/multi_gpu/cacheTransceiverTest --gtest_filter="AsymmetricCaseTest1WithCPForMLA/AsymmetricalCacheTest.TestCase/*"
$ pytest tests/integration/defs/accuracy/test_disaggregated_serving.py::TestDeepSeekV3Lite::test_auto_dtype_with_helix[nccl-cudagraph:none-pp1tp2cp2] -s -v
$ pytest tests/integration/defs/accuracy/test_disaggregated_serving.py::TestDeepSeekV3Lite::test_auto_dtype_with_helix[nccl-cudagraph:with_padding-pp1tp2cp2] -s -v
$ pytest tests/integration/defs/accuracy/test_disaggregated_serving.py::TestDeepSeekV3Lite::test_auto_dtype_with_helix[fifo-cudagraph:none-pp1tp2cp2] -s -v
$ pytest tests/integration/defs/accuracy/test_disaggregated_serving.py::TestDeepSeekV3Lite::test_auto_dtype_with_helix[fifo-cudagraph:with_padding-pp1tp2cp2] -s -v
```

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.


- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **Bug Fixes**
  * Corrected rank calculations and group mappings for proper tensor and context parallelism handling across distributed systems.
  * Fixed device mesh dimension ordering to properly reflect parallelism configuration.

* **Tests**
  * Expanded disaggregated serving test coverage with multiple parallelism configuration combinations.
  * Updated test database entries and parameterization to support new parallelism testing scenarios.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

brb-nv committed:
[TRTLLM-9465][fix] Swap TP-CP grouping order

Signed-off-by: Balaram Buddharaju <169953907+brb-nv@users.noreply.github.com>

brb-nv commented on tensorrt_llm/mapping.py:
```diff
@@ -378,17 +376,17 @@ class Mapping(MappingBase):
 
-    - [0, 1, 2, 3]
-    - [4, 5, 6, 7]
+    - [0, 1]
```
Probably needs an update.

brb-nv committed:
cleaner order of iteration

Signed-off-by: Balaram Buddharaju <169953907+brb-nv@users.noreply.github.com>

yuxianq commented on tensorrt_llm/models/modeling_utils.py:
```diff
@@ -740,10 +740,11 @@ def from_checkpoint(
             tp_size = config.mapping.tp_size
             cp_size = config.mapping.cp_size
-            rank = rank % tp_size + rank // (tp_size * cp_size) * tp_size
+            rank = (rank % (tp_size * cp_size)) // cp_size + rank // (
```
I think `rank = rank // cp_size` is enough.

tensorrt-cicd commented:
[PR_Github #30216](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/30216/) [ run ] completed with state `SUCCESS`. Commit: `872167b` 
[/LLM/main/L0_MergeRequest_PR pipeline #23259](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/23259/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

brb-nv committed:
address comments from Yuxian

Signed-off-by: Balaram Buddharaju <169953907+brb-nv@users.noreply.github.com>

brb-nv commented on tensorrt_llm/_torch/device_mesh.py:
```diff
@@ -118,15 +118,17 @@ def build_mesh(self):
         else:
-            dims += ["tp"]
-            shape += [self.tp_size]
+            dims += ["cp"]
```
Great point. Updated. https://github.com/NVIDIA/TensorRT-LLM/pull/10350/commits/fe70dd19b942569fa1ece937b4ae5ae362b8538e

tensorrt-cicd commented:
[PR_Github #30265](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/30265/) [ run ] triggered by Bot. Commit: `fe70dd1`

tensorrt-cicd commented:
[PR_Github #30285](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/30285/) [ run ] triggered by Bot. Commit: `fe70dd1`

tensorrt-cicd commented:
[PR_Github #30290](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/30290/) [ run ] triggered by Bot. Commit: `edde143`

brb-nv committed:
unconditionally add cp to device mesh

$ pytest tests/integration/defs/examples/test_ray.py::test_llm_inference_distributed_ray -s -v

Signed-off-by: Balaram Buddharaju <169953907+brb-nv@users.noreply.github.com>

tensorrt-cicd commented:
[PR_Github #30310](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/30310/) [ run ] completed with state `SUCCESS`. Commit: `edde143` 
[/LLM/main/L0_MergeRequest_PR pipeline #23344](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/23344/) completed with status: 'FAILURE'

⚠️ **Action Required:**
- Please check the failed tests and fix your PR
- If you cannot view the failures, ask the CI triggerer to share details
- Once fixed, request an NVIDIA team member to trigger CI again

brb-nv committed:
update test expectations for targetTest.CacheStateNODP

Signed-off-by: Balaram Buddharaju <169953907+brb-nv@users.noreply.github.com>

brb-nv commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #30323](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/30323/) [ run ] triggered by Bot. Commit: `7577996`

tensorrt-cicd commented:
[PR_Github #30349](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/30349/) [ run ] triggered by Bot. Commit: `896a9f3`
